src/apps/applib/include/benchmark.h

```diff
@@ -20,30 +20,51 @@
 #define BENCHMARK_H
 
+#ifdef _WIN32
+
+#include <Windows.h>
+
+#define START_TIMER                   \
+    LARGE_INTEGER start;              \
+    LARGE_INTEGER freq;               \
+    QueryPerformanceFrequency(&freq); \
+    QueryPerformanceCounter(&start)
+
+#define END_TIMER(var)             \
+    LARGE_INTEGER end;             \
+    QueryPerformanceCounter(&end); \
+    const long double var =        \
+        ((long double)(end.QuadPart - start.QuadPart)) / freq.QuadPart * 1E9
+
+#else  // !defined(_WIN32)
+
 #include <time.h>
 
-#define BEGIN_BENCHMARKS()             \
-    int main(int argc, char* argv[]) { \
-        clock_t start;                 \
-        clock_t diff;                  \
-        double duration;               \
-        int iterations;                \
-        char* name;
-
-#define BENCHMARK(NAME, ITERATIONS, BODY)                                    \
-    start = clock();                                                         \
-    iterations = ITERATIONS;                                                 \
-    name = #NAME;                                                            \
-    for (int i = 0; i < iterations; i++) {                                   \
-        BODY;                                                                \
-    }                                                                        \
-    diff = clock() - start;                                                  \
-    duration = diff * 1000 / CLOCKS_PER_SEC;                                 \
-    printf("\t-- %s: %f milliseconds per iteration (%d iterations)\n", name, \
-           duration / iterations, iterations);
-
-#define END_BENCHMARKS() \
-    ;                    \
-    }
+#define START_TIMER        \
+    struct timespec start; \
+    clock_gettime(CLOCK_MONOTONIC, &start)
+
+#define END_TIMER(var)                                         \
+    struct timespec end;                                       \
+    clock_gettime(CLOCK_MONOTONIC, &end);                      \
+    const long double var = (end.tv_sec * 1E9 + end.tv_nsec) - \
+                            (start.tv_sec * 1E9 + start.tv_nsec)
+
+#endif
+
+#define BEGIN_BENCHMARKS() int main(int argc, char* argv[]) {
+#define BENCHMARK(NAME, ITERATIONS, BODY)                               \
+    do {                                                                \
+        const int iterations = ITERATIONS;                              \
+        const char* name = #NAME;                                       \
+        START_TIMER;                                                    \
+        for (int i = 0; i < iterations; i++) {                          \
+            BODY;                                                       \
+        }                                                               \
+        END_TIMER(duration);                                            \
+        printf("\t-- %s: %Le ns per iteration (%d iterations)\n", name, \
```

Comment:
Can we change the formatting for nanoseconds here? In the example you posted on #125 I find it really hard to easily see order-of-magnitude differences. I'd prefer to still express this in ms, or maybe translate this to ops/second?

Reply:
I get that. I was considering working out a way to change the unit dynamically.